### ISSUE TITLE
yii\db\ColumnSchema::typecast() converts empty string to null even if column doesn't allow null

### ISSUE THREAD

**Issue by brandonkelly** [yiisoft/yii2#16364]
### What steps will reproduce the problem?

1. Create a MySQL database table with a `text` column.
2. Run this:

   ```php
   Yii::$app->db->createCommand()->insert('table_name', [
       'column_name' => '' // empty string
   ])->execute();
   ```

### What is the expected result?

The row should be inserted.

### What do you get instead?

A yii\db\IntegrityException exception is thrown because the empty string [gets converted to `null`](https://github.com/yiisoft/yii2/blob/01636ae45a51b3b5d4c2796b05c16e27678adb36/framework/db/ColumnSchema.php#L128) by yii\db\ColumnSchema::typecast(), despite the ColumnSchema’s `$allowNull` property being set to `false`.

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.15.1

**Comment by brandonkelly** [comment]
I fixed this in Craft by extending `yii\db\ColumnSchema` and overriding the `typecast()` method: craftcms/cms@6140530a69692a90d08e5eacb24cdfe0b984edb9

**Comment by samdark** [comment]
Would you like to make a pull request to Yii?

**Comment by brandonkelly** [comment]
done :)

**Comment by brandonkelly** [comment]
Nevermind, I misunderstood what the code was doing (missed the `!` before `in_array(`). Bug ended up being due to Craft adding support for additional text column types (`mediumtext`, etc.). So it makes sense that the fix should only be implemented in Craft: craftcms/cms@c3568e4e444cfcb0d73a0ea553cd0bae183f2d07
